SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialSender.cpp

```diff
@@ -541,36 +541,68 @@ void USpatialSender::SendRPC(TSharedRef<FPendingRPCParams> Params)
 		break;
 	}
 	case SCHEMA_NetMulticastRPC:
-	case SCHEMA_ClientUnreliableRPC:
-	case SCHEMA_ServerUnreliableRPC:
 	{
-		Worker_ComponentId ComponentId;
-		if (RPCInfo->Type == SCHEMA_ClientUnreliableRPC)
-		{
-			ComponentId = SpatialConstants::SERVER_RPC_ENDPOINT_COMPONENT_ID;
-		}
-		else if (RPCInfo->Type == SCHEMA_ServerUnreliableRPC)
+		FUnrealObjectRef TargetObjectRef(PackageMap->GetUnrealObjectRefFromNetGUID(PackageMap->GetNetGUIDFromObject(TargetObject)));
+		if (TargetObjectRef != FUnrealObjectRef::UNRESOLVED_OBJECT_REF)
 		{
-			ComponentId = SpatialConstants::CLIENT_RPC_ENDPOINT_COMPONENT_ID;
+			EntityId = TargetObjectRef.Entity;
+			Worker_ComponentId ComponentId = SpatialConstants::NETMULTICAST_RPCS_COMPONENT_ID;
+
+			Worker_ComponentUpdate ComponentUpdate = CreateUnreliableRPCUpdate(TargetObject, Params->Function, Params->Parameters.GetData(), ComponentId, RPCInfo->Index, UnresolvedObject);
+
+			if (!UnresolvedObject)
+			{
+				check(EntityId != SpatialConstants::INVALID_ENTITY_ID);
+
+				if (!NetDriver->StaticComponentView->HasAuthority(EntityId, ComponentUpdate.component_id))
+				{
+					UE_LOG(LogSpatialSender, Verbose, TEXT("Trying to send RPC %s component update but don't have authority! Update will not be sent. Entity: %lld"), *Params->Function->GetName(), EntityId);
+					return;
+				}
+
+				Connection->SendComponentUpdate(EntityId, &ComponentUpdate);
+			}
 		}
-		else
+		break;
+	}
+	case SCHEMA_ClientReliableRPC:
+	case SCHEMA_ServerReliableRPC:
+	case SCHEMA_ClientUnreliableRPC:
+	case SCHEMA_ServerUnreliableRPC:
+	{
+		FUnrealObjectRef TargetObjectRef(PackageMap->GetUnrealObjectRefFromNetGUID(PackageMap->GetNetGUIDFromObject(TargetObject)));
+		if (TargetObjectRef != FUnrealObjectRef::UNRESOLVED_OBJECT_REF)
 		{
-			ComponentId = SpatialConstants::NETMULTICAST_RPCS_COMPONENT_ID;
-		}
+			EntityId = TargetObjectRef.Entity;
+			if (NetDriver->IsEntityListening(EntityId))
+			{
+				Worker_ComponentId ComponentId;
+				if ((RPCInfo->Type == SCHEMA_ClientUnreliableRPC) ||
+					(RPCInfo->Type == SCHEMA_ClientReliableRPC))
+				{
+					ComponentId = SpatialConstants::SERVER_RPC_ENDPOINT_COMPONENT_ID;
+				}
+				else if ((RPCInfo->Type == SCHEMA_ServerUnreliableRPC) ||
+					(RPCInfo->Type == SCHEMA_ServerReliableRPC))
+				{
+					ComponentId = SpatialConstants::CLIENT_RPC_ENDPOINT_COMPONENT_ID;
+				}
 
-		Worker_ComponentUpdate ComponentUpdate = CreateUnreliableRPCUpdate(TargetObject, Params->Function, Params->Parameters.GetData(), ComponentId, RPCInfo->Index, EntityId, UnresolvedObject);
+				Worker_ComponentUpdate ComponentUpdate = CreateUnreliableRPCUpdate(TargetObject, Params->Function, Params->Parameters.GetData(), ComponentId, RPCInfo->Index, UnresolvedObject);
```

Comment:
I'd change this name as it's no longer just unreliable rpcs. IMO the internal functions should probably just say what consistency they guarantee or have some other name - like event rpc. However this is not a big deal as this will likely change a bit again in the near future.

Reply:
What about `CreateRPCEventUpdate`?

Reply:
Renamed to `CreateRPCEventUpdate`